src/Xamarin.Android.Build.Tasks/Utilities/XAAssemblyResolver.cs

```diff
@@ -0,0 +1,333 @@
+			ReadSymbols                     = haveDebugSymbols,
+			ReadWrite                       = readerParameters.ReadWrite,
+			ReflectionImporterProvider      = readerParameters.ReflectionImporterProvider,
+			SymbolReaderProvider            = readerParameters.SymbolReaderProvider,
+			SymbolStream                    = readerParameters.SymbolStream,
+		};
+		try {
+			return LoadFromMemoryMappedFile (filePath, loadReaderParams);
+		} catch (Exception ex) {
+			log.LogWarning ($"Failed to read '{filePath}' with debugging symbols. Retrying to load it without it. Error details are logged below.");
+			log.LogWarning ($"{ex.ToString ()}");
+			loadReaderParams.ReadSymbols = false;
+			return LoadFromMemoryMappedFile (filePath, loadReaderParams);
+		}
+	}
+
+	AssemblyDefinition LoadFromMemoryMappedFile (string file, ReaderParameters options)
+	{
+		// We can't use MemoryMappedFile when ReadWrite is true
+		if (options.ReadWrite) {
+			return AssemblyDefinition.ReadAssembly (file, options);
+		}
+
+		bool origReadSymbols = options.ReadSymbols;
+		MemoryMappedViewStream? viewStream = null;
+		try {
+			// We must disable reading of symbols, even if they were present, because Cecil is unable to find the symbols file when
+			// assembly file name is unknown, and this is precisely the case when reading module from a stream.
+			// Until this issue is resolved, skipping symbol read saves time because reading exception isn't thrown and we don't
+			// retry the load.
+			options.ReadSymbols = false;
+
+			// Create stream because CreateFromFile(string, ...) uses FileShare.None which is too strict
+			using var fileStream = new FileStream (file, FileMode.Open, FileAccess.Read, FileShare.Read, 4096, false);
+			using var mappedFile = MemoryMappedFile.CreateFromFile (fileStream, null, fileStream.Length, MemoryMappedFileAccess.Read, HandleInheritability.None, true);
+			viewStream = mappedFile.CreateViewStream (0, 0, MemoryMappedFileAccess.Read);
+
+			AssemblyDefinition result = ModuleDefinition.ReadModule (viewStream, options).Assembly;
+			viewStreams.Add (viewStream);
+
+			// We transferred the ownership of the viewStream to the collection.
+			viewStream = null;
+
+			return result;
+		} finally {
+			options.ReadSymbols = origReadSymbols;
+			viewStream?.Dispose ();
+		}
+	}
+
+	AssemblyDefinition? SelectAssembly (AndroidTargetArch arch, string assemblyName, CacheEntry? entry, bool loading)
+	{
+		if (entry == null) {
+			// Should "never" happen...
+			throw new ArgumentNullException (nameof (entry));
+		}
+
+		if (arch == AndroidTargetArch.None) {
+			// Disabled for now, generates too much noise.
+			// if (entry.Assemblies.Count > 1) {
+			// 	log.LogWarning ($"Architecture-agnostic entry requested for architecture-specific assembly '{assemblyName}'");
+			// }
+			return entry.Default;
+		}
+
+		if (!entry.Assemblies.TryGetValue (arch, out AssemblyDefinition? asm)) {
+			if (loading) {
```

Comment:
I think the `loading` parameter should be renamed for clarity, because I don't understand why `loading:true` would result in no assembly being returned.  Additionally, `loading:true` is used by "force load" path, which doesn't appear to be used, so we have additional unused codepaths.

I don't think we needed `bool loading` at all.

Reply:
If we don't have `loading` then for every per-RID assembly other than the first one we'll get the "assembly not found" exception.

Reply:
This code:

```csharp
	public virtual AssemblyDefinition? Load (AndroidTargetArch arch, string filePath, ReaderParameters? readerParameters = null)
	{
		string name = Path.GetFileNameWithoutExtension (filePath);
		AssemblyDefinition? assembly;
		if (cache.TryGetValue (name, out CacheEntry? entry)) {
			assembly = SelectAssembly (arch, name, entry, loading: true);
			if (assembly != null) {
				return assembly;
			}
		}
```

will be called for every per-RID assembly, with the same name but different path/RID. If `loading` is absent, `SelectAssembly` will throw because it won't find an entry for `arch`